pkg/handler/enqueue_owner.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var _ EventHandler = &EnqueueOwner{}
+
+var log = logf.KBLog.WithName("eventhandler").WithName("EnqueueOwner")
+
+// EnqueueOwner enqueues Requests for the Owners of an object.  E.g. an object that created
+// another object.
+//
+// If a ReplicaSet creates Pods, reconcile the ReplicaSet in response to events on Pods that it created using:
+//
+// - a KindSource with Type Pod.
+//
+// - a EnqueueOwner with OwnerType ReplicaSet.
+type EnqueueOwner struct {
+	// OwnerType is the type of the Owner object to look for in OwnerReferences.  Only Group and Kind are compared.
+	OwnerType runtime.Object
+
+	// IsController if set will only look at the first OwnerReference with Controller: true.
+	IsController bool
+
+	// groupKind is the cached Group and Kind from OwnerType
+	groupKind schema.GroupKind
+}
+
+var _ inject.Scheme = &EnqueueOwner{}
+
+// InjectScheme is called by the Controller to provide a singleton scheme to the EnqueueOwner.
+func (e *EnqueueOwner) InjectScheme(s *runtime.Scheme) error {
+	return e.parseOwnerTypeGroupKind(s)
+}
```

Comment:
would suggest moving `InjectScheme` related code at the bottom from code readability perspective, sort of coming in the way :)